### PR TITLE
Formatter refactoring

### PR DESCRIPTION
**Pull request by PolinaShneider**
Now `Formatter` instead of `TextEdits` array returns fully formatted config text, which provides better editing flexibility.

**Note:** document is not being literally formatted, but rather recreated line by line according to formatting rules.

## New features:
1. Add blank line before new section (except for `configuration`)
2. Remove extra blank lines between settings
3. Remove trailing spaces
4. Insert two blank lines at the end of the document

## Demo:
blankLinesFormatting

### CODE REVIEW

File: src/formatter.ts [hunk 1]
```diff
@@ -1,17 +1,31 @@
...
+import { isEmpty } from "./util";
 
 interface Section {
     indent?: string;
     name?: string;
 }
+
+/** Extended formatting options, supporting blank lines formatting possibility */
+export interface ExtendedFormattingOptions extends FormattingOptions {
+    blankLinesAtEnd?: number;
```
Comment (annstriganova) [inline src/formatter.ts]: Does we actually need this option? Shouldn't it be always `blankLinesAtEnd=3`? If we use it only in tests, I suggest to fix tests.

Comment (PolinaShneider) [reply src/formatter.ts]: I thought, this option should be configurable. We might not always want exactly 3 lines. If a kind of hardcoding 3 lines is okay, I'll do it.

Comment (annstriganova) [reply src/formatter.ts]: Since it's necessary to ensure correct scroll in ChartLab (6474), let's hardcode.

File: src/formatter.ts [hunk 2]
```diff
@@ -68,43 +83,38 @@ export class Formatter {
...
-        this.lines = text.split("\n");
     }
 
     /**
      * Reads the document line by line and calls corresponding formatting functions
-     * @returns array of text edits to properly format document
+     * @returns array containing single text edit with fully formatted document
      */
-    public lineByLine(): TextEdit[] {
+    public lineByLine(text: string): TextEdit[] {
```
Comment (VeselovAlex) [inline src/formatter.ts]: 1. Rename it, `lineByLine` is not a best name for public parsing method
2. Just resurn text and let clients to decide what to do with formatted content, hardcoding document positions is not a good idea

Comment (VeselovAlex) [reply src/formatter.ts]: For future releases: investigate, can we separate parsing and formatting stages. Using own hardcoded parsers for both validator/formatter is slow, hard to support and possibly buggy

File: src/formatter.ts [hunk 3]
```diff
@@ -133,11 +187,11 @@ export class Formatter {
         if (!buffer.length) {
             return;
         }
-        const content = buffer.join("\n");
+        const unformattedCode = buffer.join("\n");
 
         try {
             /** Parse and format JavaScript */
-            const parsedCode = parseScript(content);
+            const parsedCode = parseScript(unformattedCode);
```
Comment (VeselovAlex) [inline src/formatter.ts]: Extract JS formatting functionality to their own class. Create a factory, which will return nested content formatter based on language (JS/SQL/CSV). Formatters for langauges other than JS can be implemented later

File: src/formatter.ts [hunk 4]
```diff
@@ -1,17 +1,51 @@
...
+    tabSize: 2
+};
+
+/**
+ * Language formatting configuration
+ * languageId - language id
+ * endRegex - regex specifying that language code block finished
+ * getOptions - function computing language formatting options based on current indent
+ */
+interface languageConfiguration {
```
Comment (VeselovAlex) [inline src/formatter.ts]: Use PascalCase for interfaces

File: src/formatter.ts [hunk 5]
```diff
@@ -1,17 +1,51 @@
...
+    languageId: string,
+    endRegex: RegExp,
+    getOptions(indent?: string, tabSize?: number): LanguageFormattingOptions
```
Comment (VeselovAlex) [inline src/formatter.ts]: IMHO the arguments should not be optional.

**Pull request merged**